src/aspire/source/coordinates.py

```diff
@@ -273,170 +300,119 @@ def _get_mrc_shapes(self):
+            `CoordinateSource`.
         """
-        if not len(ctf_files) == len(self.mrc_paths):
+        relion_version = getRelionStarFileVersion(ctf)
+        if not relion_version:
             raise ValueError(
-                "Number of CTF STAR files must match number of micrographs."
+                f"Cannot recognize {ctf} as a valid Relion STAR file containing micrograph information."
+            )
+        if relion_version == "3.0":
+            data_block = RelionLegacyDataStarFile(ctf).get_block_by_index(0)
+        if relion_version == "3.1":
+            starfile = RelionDataStarFile(ctf)
+            # populate CTF parameters in main data block (from optics groups block)
+            # specifically: voltage, Cs, amplitude_contrast, micrograph pixel size
+            # this feature does not exist for 3.0 starfile
+            data_block = starfile.apply_optics_block()
+
+        # data_block is a pandas Dataframe containing the micrographs
+        if not len(data_block) == self.num_micrographs:
+            raise ValueError(
+                f"{starfile.filepath} has CTF information for {len(data_block)}",
+                f" micrographs but this source has {self.num_micrographs} micrographs.",
             )
 
-        for mrc_idx, ctf_file in enumerate(ctf_files):
-            params = self._read_ctf_star(ctf_file)
-            self._process_each_ctf(params, mrc_idx)
+        self._extract_ctf(data_block)
 
-    def _process_each_ctf(self, params, mrc_idx):
+    def _extract_ctf(self, df):
         """
-        Given a unique set of CTF parameters, create a CTFFilter and populate source metadata.
-        :param params: Dictionary of CTF parameters.
-        :param mrc_idx: Index of the micrograph corresponding to these CTF parameters.
+        Receives a flattened DataFrame containing micrograph CTF information, and populates
+            the Source's CTF Filters, filter indices, and metadata.
         """
-        # find particle indices corresponding to this micrograph
-        indices = [
-            idx for idx, particle in enumerate(self.particles) if particle[0] == mrc_idx
+        # required CTF params excluding pixel size
+        CTF_params = [
+            "_rlnVoltage",
+            "_rlnDefocusU",
+            "_rlnDefocusV",
+            "_rlnDefocusAngle",
+            "_rlnSphericalAberration",
+            "_rlnAmplitudeContrast",
+            "_rlnMicrographPixelSize",
         ]
-        # add CTF filter to unique filters
-        self._unique_filters.append(
+
+        # get unique ctfs from the data block
+        # i'th entry of `indices` contains the index of `filter_params` with corresponding CTF params
+        filter_params, indices = np.unique(
+            df[CTF_params].astype(self.dtype).values, return_inverse=True, axis=0
+        )
+
+        # convert defocus_ang from degrees to radians
+        filter_params[:, 3] *= np.pi / 180.0
```

Comment:
was this done somewhere else before, or just missing?

Reply:
This was missing before (the metadata test in `test_coordinate_source` has been updated to match).

It now matches `RelionSource`:

https://github.com/ComputationalCryoEM/ASPIRE-Python/blob/1554bbba7f0c77b5c6756fdca8c1ec61351b1b8a/src/aspire/source/relion.py#L151-L163

Reply:
Cool.